ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -293,71 +302,178 @@ private module Cached {
 
 import Cached
 
-private DataFlow::LocalSourceNode trackInstance(Module tp, TypeTracker t) {
+/**
+ * Holds if `self` variable has module type `m`.
+ *
+ * The Boolean `exact` indicates whether `self` has exactly the type `m`
+ * (and not possibly a sub class of `m`).
+ */
+pragma[nomagic]
+private predicate selfHasType(SelfVariable self, boolean exact, Module m) {
+  exists(Scope scope | scope = self.getDeclaringScope() |
+    // `self` in module
+    (
+      if scope instanceof Toplevel
+      then m = TResolved("Object") and exact = true
+      else (
+        m = scope.(ModuleBase).getModule() and
+        exact = false
+      )
+    )
+    or
+    // `self` in method
+    exists(ModuleBase encl |
+      encl = scope.(MethodBase).getEnclosingModule() and
+      exact = false and
+      if encl instanceof SingletonClass
+      then m = encl.getEnclosingModule().getModule()
+      else m = encl.getModule()
+    )
+  )
+}
+
+/**
+ * Holds if `self` SSA node has module type `m`.
+ *
+ * The Boolean `exact` indicates whether `self` has exactly the type `m`
+ * (and not possibly a sub class of `m`).
+ */
+private predicate selfNodeHasType(SsaSelfDefinitionNode self, boolean exact, Module m) {
+  selfHasType(self.getVariable(), exact, m)
+}
+
+/**
+ * Holds if expression `e` refers to module `m`.
+ *
+ * The Boolean `exact` indicates whether `n` refers to exactly `m` (and not
+ * possibly a sub class).
+ */
+private predicate resolveModule(Expr e, boolean exact, Module m) {
+  m = resolveConstantReadAccess(e) and
+  exact = true
+  or
+  selfHasType(e.(SelfVariableReadAccess).getVariable(), exact, m)
+}
+
+/**
+ * Holds if node `n` refers to module `m`.
+ *
+ * The Boolean `exact` indicates whether `n` refers to exactly `m` (and not
+ * possibly a sub class).
+ */
+private predicate resolveModuleNode(DataFlow::Node n, boolean exact, Module m) {
+  m = resolveConstantReadAccess(n.asExpr().getExpr()) and
+  exact = true
+  or
+  selfNodeHasType(n, exact, m)
+}
+
+private DataFlow::LocalSourceNode trackInstance(Module tp, boolean exact, TypeTracker t) {
   t.start() and
   (
-    result.asExpr().getExpr() instanceof NilLiteral and tp = TResolved("NilClass")
+    result.asExpr().getExpr() instanceof NilLiteral and
+    tp = TResolved("NilClass") and
+    exact = false
```

Comment:
Was this supposed to say `exact = true` here?

Reply:
I guess so, yes.